scaaml/stats/print_stats.py

```diff
@@ -0,0 +1,146 @@
+standalone."""
+
+from itertools import product
+from typing import Dict
+
+from tabulate import tabulate
+from tqdm import tqdm
+
+from scaaml.io import Dataset
+from .ap_counter import APCounter
+from .ap_checker import APChecker
+from .example_iterator import ExampleIterator
+from .trace_stddev_of_avg import STDDEVofAVGofTraces
+
+
+class PrintStats:
+    """Gather and print warnings from attack point checks and trace
+    statistics.
+
+    Example use:
+
+    # On an existing dataset:
+    print_stats = PrintStats.from_config(dataset_path='/mnt/path/to/dataset')
+
+    # During capture:
+    print_stats = PrintStats(measurements_info, attack_points_info)
+    for example in example_iterator:
+        print_stats.add_example(example=example,
+                                split=split,
+                                group=group,
+                                part=part)
+    print_stats.print()
+    """
+    def __init__(self, measurements_info: Dict,
+                 attack_points_info: Dict) -> None:
+        self._measurements_info = measurements_info
+        self._attack_point_info = attack_points_info
+        self._split_ap_counters = {
+            split: {
+                ap_name: [
+                    APCounter(attack_point_info=ap_info),
+                ]
+                for ap_name, ap_info in attack_points_info.items()
+            }
+            for split in Dataset.SPLITS
+        }
+        self._split_trace_stats = {
+            split: {
+                trace_name: [
+                    STDDEVofAVGofTraces(),
+                ]
+                for trace_name in measurements_info
+            }
+            for split in Dataset.SPLITS
+        }
+
+    @staticmethod
+    def from_config(dataset_path: str) -> None:
+        """Read a dataset and print all attack point check warnings and trace
+        statistics.
+
+        Args:
+          dataset_path: Root path of the dataset.
+        """
+        dataset = Dataset.from_config(dataset_path)
+        print_stats = PrintStats(measurements_info=dataset.measurements_info,
+                                 attack_points_info=dataset.attack_points_info)
+        all_splits = set(dataset.shards_list.keys())
+        all_groups = set(shard['group'] for split in Dataset.SPLITS
+                         for shard in dataset.shards_list[split])
+        all_parts = set(shard['part'] for split in Dataset.SPLITS
+                        for shard in dataset.shards_list[split])
+        prod = list(product(all_splits, all_groups, all_parts))
+        for split, group, part in tqdm(prod):
+            print(f'split: {split} group: {group} part: {part}')
+            example_iterator = ExampleIterator(dataset_path=dataset_path,
+                                               split=split,
+                                               group=group,
+                                               part=part)
+            for example in tqdm(example_iterator):
```

Comment:
Having a tqdm loop within a tqdm loop will likely cause artifacts to be left on the screen (that is: do you see a bunch of progress bars?).

If you change the innermost loop to: `for example in tqdm(example_iterator, leave=False):` the issue should go away.